File: haskell/bob/Bob.hs
module Bob where

import Data.Char (isAlpha, isUpper, isSpace)

isQuestion xs = last xs == '?'
isYelling xs = any isAlpha xs && all isUpper letters
isBlank = all isSpace

responseFor x
  | isBlank x    = "Fine. Be that way!"
  | isYelling x  = "Whoa, chill out!"
  | isQuestion x = "Sure."
  | otherwise    = "Whatever."

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :refer [blank?]]))

(defn yelling? [phrase]
  (let [letters (filter #(Character/isLetter %) phrase)]
    (if (seq letters)
      (every? #(Character/isUpperCase %) letters)
      false)))

(defn question? [phrase]
  (= (last phrase) \?))

(defn response-for [phrase]
  (cond
    (blank? phrase) "Fine. Be that way!"
    (yelling? phrase) "Whoa, chill out!"
    (question? phrase) "Sure."
    :else "Whatever."))